cmd/orch-tester/orch_tester.go

```diff
@@ -549,3 +603,219 @@ func (s *streamerClient) getOrchestrators() ([]*orch, error) {
 
 	return data.Data.Transcoders, nil
 }
+
+// Get a random source/rendition sample segment and write it to external storage
+func (s *streamerClient) randomSample(ctx context.Context, mid string, orch string) error {
+	randSeqNo := rand.Int() % numSegments
+	randProfile := s.profiles[rand.Int()%len(s.profiles)]
+	//stream/manifestID/profile/SeqNo.ts
+	sourceUrl := fmt.Sprintf("%v/stream/%v_0_0/source/%v.ts", s.broadcaster, mid, randSeqNo)
+	renditionUrl := fmt.Sprintf("%v/stream/%v_0_0/%v/%v.ts", s.broadcaster, mid, randProfile, randSeqNo)
+	var (
+		rendition []byte
+		source    []byte
+		err       error
+	)
+	tick := time.NewTicker(2 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			glog.Info("random sampler timed out")
+			// If we haven't found the segment try to sample from whatever playlist we can
+			rendition, source, renditionUrl, err = s.sampleFromPlaylist(mid)
+			if err != nil {
+				glog.Error(err)
+			}
+			break
+		case <-tick.C:
+			rendition, err = s.downloadSegment(renditionUrl)
+			if err != nil {
+				glog.V(model.DEBUG).Info(err)
+				continue
+			}
+			source, err = s.downloadSegment(sourceUrl)
+			if err != nil {
+				glog.V(model.DEBUG).Info(err)
+				continue
+			}
+			break
+		}
+		break
+	}
+
+	if len(rendition) == 0 || len(source) == 0 {
+		return fmt.Errorf("no segments found")
+	}
+
+	urlSplit := strings.Split(renditionUrl, "/")
+	rendS := urlSplit[len(urlSplit)-2]
+	fname := urlSplit[len(urlSplit)-1]
+
+	rendF := fileName(orch, mid, rendS, fname)
+	sourceF := fileName(orch, mid, "source", fname)
+
+	src, _, err := testers.SaveToExternalStorage(sourceF, source)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	glog.Infof("Wrote source segment to storage url=%v", src)
+
+	rend, _, err := testers.SaveToExternalStorage(rendF, rendition)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	glog.Infof("Wrote rendition to storage url=%v", rend)
+	return nil
+}
+
+func baseFileName() string {
+	y, m, d := start.Date()
+	return fmt.Sprintf("%v-%v-%v", y, m, d)
+}
+
+func fileName(orch, mid, rend, fname string) string {
+	return fmt.Sprintf("%v/%v/%v-%v-%v", baseFileName(), orch, mid, rend, fname)
+}
+
+func (s *streamerClient) sampleFromPlaylist(mid string) (source, rendition []byte, url string, err error) {
+	mpl, err := s.downloadMasterPlaylist(mid)
+	if err != nil {
+		return nil, nil, "", err
+	}
+
+	sourceURI := fmt.Sprintf("%v_0_0/source.m3u8", mid)
+	sourcePl, err := s.downloadMediaPlaylist(sourceURI)
+	if err != nil {
+		return nil, nil, "", err
+	}
+	renditionPl, err := s.downloadMediaPlaylist(mpl.Variants[rand.Int()%len(mpl.Variants)].URI)
+	if err != nil {
+		return nil, nil, "", err
+	}
+
+	if len(renditionPl.Segments) == 0 {
+		return nil, nil, "", fmt.Errorf("no transcoded segments")
+	}
+
+	var rendSegs []*m3u8.MediaSegment
+	var sourceSegs []*m3u8.MediaSegment
+	for _, v := range renditionPl.Segments {
```

Comment:
Since we can fetch the length of the playlist via the `Len()` method as mentioned in this [comment](https://github.com/livepeer/stream-tester/pull/68#discussion_r546942815) by @darkdarkdragon, I think we can omit these loops and make the following update on L721:

```
segIdx := r % int(math.Min(float64(renditionPl.Len()), float64(sourcePl.Len()))
```

which should ensure that we never get a nil segment (that might be at the end of the slice stored by the playlist) when indexing into the Segments slice.